[MM-33806, MM-34097] Detox/E2E: Sample iOS E2E for certificate auth

josephbaylon opened the pull request:
#### Summary
- Added e2e for secure api client (`get/put/patch/post/delete`), file client (`upload/fast image`),  and websocket client
- Added new detox screen for APIClientImportP12
- Removed third-party mockserver; created in-house mockserver
- Added secure capability to mockserver, file server, and websocket server
- Added ability to download binary file from url and write binary file to native
- Added new testIDs and helper functions

#### Ticket Link
JIRA:
- https://mattermost.atlassian.net/browse/MM-33806
- https://mattermost.atlassian.net/browse/MM-34097

Fixes merged in this PR
- https://mattermost.atlassian.net/browse/MM-34420
- https://mattermost.atlassian.net/browse/MM-34374
- https://mattermost.atlassian.net/browse/MM-34218
- https://mattermost.atlassian.net/browse/MM-34612
- https://mattermost.atlassian.net/browse/MM-34422

![Screen Shot 2021-04-08 at 9 00 38 AM](https://user-images.githubusercontent.com/487991/114059311-19557e80-9849-11eb-95e5-535a72a1fd66.png)

josephbaylon commented:
Hey @migbot I created a secure server for manual testing. It works when I use postman with self-signed cert, however I couldn't get it to work on the example app. If you have some time, can you please take a look at the changes to see why it wouldn't work?

I'm setting the trust self-signed here https://github.com/mattermost/react-native-network-client/pull/57/files#diff-39142aa799dc871f61b7b2c2fa695ba32c4ce913018a01815b6fd810e24415d1R209

Steps,
1. run secure server `npm run secure-server`
2. bring up example app, select `Secure Server API` client
3. use `/get` | `/post` | `/put` | `/patch` | `/delete` path
4. make request

Here's the equivalent postman request against the secure server with self signed cert
![Screen Shot 2021-03-24 at 5 10 05 PM](https://user-images.githubusercontent.com/487991/112400049-b39fb900-8cc4-11eb-91ba-2b7620b7e235.png)

josephbaylon commented:
Here are the steps I took to generate the server certs and client certs

1. Generate certs using https://github.com/julie-ng/nodejs-certificate-auth
```
./generate-certs.sh
```
2. Used the generated server certs in the secure server I created
```
example/certs/server_cert.pem
example/certs/server_key.pem
example/secure_server.js
```
3. Converted the client cert from pem to p12 and added `password` as password
```
openssl pkcs12 -export -out alice_cert.p12 -in alice_cert.pem -inkey alice_key.pem
```
4. Renamed alice_cert.p12 to client_cert.p12 and uploaded to my iCloud drive so I can access from my simulator
```
example/certs/client_cert.p12
```
5. Ran secure server
```
npm run secure-server
```
6. Ran the example app, import the p12 cert with default `password`
7. Perform a get request

![Screen Shot 2021-03-24 at 3 02 08 PM](https://user-images.githubusercontent.com/487991/112498280-9958f000-8d43-11eb-979e-c028f8faea54.png)

enahum commented:
@josephbaylon can you try generating the certs with this steps https://docs.mattermost.com/deployment/ssl-client-certificate.html#set-up-mutual-tls-authentication-for-the-web-app ?

enahum commented:
Also chances are that self-sign certificates for the server are not allowed. @migbot @shazm did we add a config flag to trust self sign certs into the API/Generic clients ?

enahum commented on example/servers/secure_server.js:
```diff
@@ -0,0 +1,57 @@
+
+    // Create router
+    const router = express.Router();
+    router.get("/get", authHandler);
+    router.post("/post", authHandler);
+    router.put("/put", authHandler);
+    router.patch("/patch", authHandler);
```
We should also add routes for uploads and everything else, we need to make sure the cert works with all options. Also test WS connections with certs

josephbaylon commented:
@enahum Turns out the secure server response was expected to be json so I added the fix. Self signed should work locally now. @migbot and I had a zoom call and verified that it works on his machine.

I'm having some local issues which I need to fix first in order to proceed, which might take some time. I'll address early feedback and continue adding fixes to this PR before I open for review.

josephbaylon commented on example/src/components/P12Inputs.tsx:
```diff
@@ -51,11 +52,25 @@ const P12Inputs = (props: P12InputsProps) => {
 
+    const attachCertificate = async () => {
+        const fromUrl =
+            "https://github.com/mattermost/react-native-network-client/raw/e2e-cert/example/certs/client_cert.p12";
```
I will update this url later to `https://github.com/mattermost/react-native-network-client/raw/master/example/certs/client_cert.p12` once merged to master

josephbaylon commented:
@enahum @migbot `/get | /post | /put | /patch | /delete` calls over ssl now work using secure mockserver below and preset client `Secure Mockserver API`
```
npm run secure-mockserver
```

However for stream file upload over ssl, the example app throws an timeout error. Not sure if it's a file server issue or something else. If you guys can review the code here for file server `example/servers/file_server.js` and point out the issue if there's one, that would be great. Use preset client `Secure File Upload Server API`
```
npm run secure-file-upload-server
```

As for websocket over ssl, there's a new jira ticket to implement `trustSelfSignedServerCertificate `
https://mattermost.atlassian.net/browse/MM-34374

shazm commented:
> However for stream file upload over ssl, the example app throws an timeout error. Not sure if it's a file server issue or something else. If you guys can review the code here for file server `example/servers/file_server.js` and point out the issue if there's one, that would be great. 

I couldn't pin-point the exact issue, but I wrapped the method(s) in a try/catch and noticed it was erroring out without a message so wasn't sure - might be worth updating the streamHandler to something similar?

```
    const multipartUploadHandler = (req, res, next) => {
         res.set("server", "file-server");

         if (!req.files || Object.keys(req.files).length === 0) {
             return res.status(400).send("No files were uploaded.");
         }

         console.log("Attempt to upload files...", Object.values(req.files));
         const filenames = [];

         try {
             Object.values(req.files).map((file) => {
                 const filename = file.name;
                 const filePath = `${uploadPath}/${filename}`;
                 filenames.push(filename);
                 console.log(`Uploading file: ${filename}`);

                 // Move to file path
                 file.mv(filePath, (err) => {
                     if (err) {
                         console.log(err);
                         return res.status(500).send(err);
                     }
                     console.log(`Finished! Uploaded to: ${filePath}`);
                 });
             });
         } catch (e) {
             console.log("Error", e);
         }
         console.log("Finished uploading files!", filenames);
         return res.status(200).json(filenames);
     };
```

josephbaylon commented:
@shazm Thanks for the tip! I added some more logging and it was a routing and serving issue with the static handler for fast-image that was causing some conflicts. I had to be explicit with the routing methods for file upload handler and pass the params to static handler.

I'll add more tests later and see if there are more issues I'll encounter.

josephbaylon commented:
@migbot For Import P12, when I import the p12 file multiple times on one API client, say `Secure Fast Image Server API`, it works fine. However, when I then try to import to a different API client, say `Secure File Upload Image Server API`, then I'm presented with the `APIClient-Error` upon import.

It doesn't matter what the sequence is as long as I try to import p12 to more than one secure client, the error shows. Would you happen to know why this is?

Reason I'm asking is this is preventing the test suite to run without one of the next secure client cases to fail because an p12 import already happened in a previous case.

Steps,
1. Import p12 to a secure client A (no need to make a request, just go back to client list after)
2. Attempt to import p12 to a secure client B

Expected: Successful import on secure client B
Actual: `APIClient-Error` is thrown

josephbaylon commented:
@migbot Here's a separate issue for secure fast image client https://mattermost.atlassian.net/browse/MM-34422

josephbaylon commented on example/detox/e2e/support/test_config.js:
```diff
@@ -12,6 +13,33 @@ module.exports = {
     fileUploadSiteUrl:
         process.env.FILE_UPLOAD_SITE_URL || "http://localhost:8008",
+    secureFastImageServerClientCertUrl:
+        "https://github.com/mattermost/react-native-network-client/raw/e2e-cert/example/certs/secure_fast_image_server_client_cert.p12",
```
I will update the secure client cert URLs later to https://github.com/mattermost/react-native-network-client/raw/master/example/certs/*_client_cert.p12 once merged to master

josephbaylon commented:
@migbot As discussed, I copied your changes from https://github.com/mattermost/react-native-network-client/pull/58 to this PR.

1. For WebSocket client request, the p12 import is missing (only available on creation but not when a client already exists)
2. Aside from secure fast image request, I will need to do some further testing, but looks promising.

josephbaylon commented:
@migbot I copied the latest from the PR and modified the websocket e2e to create a client with secure URL. The secure websocket tests are passing now. Only the fast image issue remains.

josephbaylon commented:
@migbot I will do the same with PR https://github.com/mattermost/react-native-network-client/pull/56, i.e. copy the fix and test in this PR.

josephbaylon commented:
@migbot For multipart, I copied the fixes from pr-56 and updated the upload test cases. For pr-58, for last commit, I only applied rejection to websocket client but not to api client so that tests can pass.

saturninoabril commented:
Sorry for the delay. Will review first thing tomorrow.

josephbaylon commented:
@saturninoabril no rush. I'm waiting on https://github.com/mattermost/react-native-network-client/pull/58 to be merged anyway so I can pull those into my PR. I added a `Do Not Merge/Awaiting PR` label.

josephbaylon commented:
I've merged Miguel's fixes into this PR and ran the suite for iOS. The secure fast image server is still an issue at least on my local (https://mattermost.atlassian.net/browse/MM-34422) but that can be on a separate PR.

@enahum @migbot @shazm @saturninoabril Please review. Thanks.

enahum commented:
What is the fast image server? And what is the issue?

josephbaylon commented:
@enahum It's just a simple server serving images, in this case static images from a file server. The issue I'm having is with the secure call of the of the image url, the image is not displayed in the client whereas with non-secure, it does. I'm not sure if it's a client issue or a server issue, that's why I'm ok with not waiting for the fix. If @migbot or @shazm has a way to test it properly, then I'll go with their testing.

saturninoabril reviewed (commented):
Could you attach test results of your local run both in iOS and Android?

In my case, all were failing except on iOS which got 2/17 spec passing only.
![Screen Shot 2021-04-08 at 7 19 19 PM](https://user-images.githubusercontent.com/5334504/114018968-6f094580-98a0-11eb-8837-6c095bd00aa6.png)

I can see logs that servers were running but I might totally missed an obvious step to run the tests successfully.

![Screen Shot 2021-04-08 at 7 28 09 PM](https://user-images.githubusercontent.com/5334504/114019221-b5f73b00-98a0-11eb-8537-6dd70a60c703.png)

saturninoabril commented on example/detox/e2e/init.js:
```diff
@@ -33,59 +52,106 @@ beforeAll(async () => {
+        } else {
+            console.log(error);
+            process.exit(1);
+        }
```
There are now too many servers. Could it be grouped into logical types, like secure and non-secure server only? In short, can it be consolidated into 2 servers only?

josephbaylon replied:
See comment https://github.com/mattermost/react-native-network-client/pull/57#issuecomment-815842308

saturninoabril commented on example/detox/e2e/test/client/create_api_client.e2e.js:
```diff
@@ -32,9 +37,73 @@ describe("Create API Client", () => {
         retryInterval: `${getRandomInt(5) + 4}`,
     };
-    const { clientListScrollView, removeClientWithName } = ClientListScreen;
+
```
Now that there are many kinds of servers. Would it make sense to add service requirement per spec file for ease of troubleshooting? Something like `await apiRequire[ServerType]`?

josephbaylon replied:
See comment https://github.com/mattermost/react-native-network-client/pull/57#issuecomment-815842308

josephbaylon commented:
@saturninoabril I can take a look at those suggestions as improvements on a separate PR. Right now, this PR is only for iOS. All servers are just launched once and is checked if they are still listening for every test file. Android will fail some tests because some of the features are not yet in order.

As for running the test, I posted the results in the description for iOS. For the failing test, can you run the failing test by itself and see the log failure?

josephbaylon commented:
I filed this ticket for me to look into later https://mattermost.atlassian.net/browse/MM-34692

josephbaylon commented:
@saturninoabril I added a test sequencer. Not sure if that will help your test runs locally. I also updated the screenshot in the description to reflect the ordered test run.

saturninoabril reviewed (approved):
Thanks Joseph! LGTM.

josephbaylon commented:
@enahum @migbot @shazm I've merged the last fix that Miguel had for iOS. This should have all the latest code and tests. I want this to be merged sooner than later so that new PRs can take advantage of the tests here. Thanks!

enahum reviewed (approved):
go go go

enahum commented:
@josephbaylon feel free to merge. two dev reviews passed
